### DISCUSSION TITLE
Resolved: `primitives` with `unstable` feature

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#3464] [General]
At any stage did we discuss the idea of adding an `unstable` feature to `primitives`? Would we then be able to do API breaking changes to anything feature guarded by `unstable` (ignoring semver)?

This would for example allow:

A. Stop using extension traits all together and just put all that stuff behind `unstable`
B. Put some abstraction leaking API's behind unstable (eg it seems to me to be non-trivial to implement `Decodable` for `Witness` without either introducing extra allocations or adding a constructor that leaks the inner abstraction)

**Comment by apoelstra** [comment]
I don't know if we discussed this. I really don't want to do this. It's super annoying to have `--all-features` enable broken stuff, and also if we had an `unstable` feature in primitives this would need to propagate to every downstream user.

**Reply by tcharding** [reply]
Humour me (or ignore me because I woke up at 2 this morning):

- `rust-bitcoin` depends on `primitives` with `unstable` unconditionally enabled - gets all functionality and is not broken with `--all-features`.
- Direct downstream users of `primitives` in theory should not enable `unstable` because doing so means they need stuff that we would have left in `bitcoin` if/when doing the extension thing anyways so either way `primitives` is useless to them.
- We release `primitives 1.0` and say "semver compliance is only maintained if you do not enable `unstable`"
- We chip away at removing `unstable` feature guards function by function, this is functionally the same as moving functions from an extension trait over to `primitives` (removing feature guard adds code to the non-`unstable` semver compliant build.

The only problem I see is it will be easier to make a mistake and not have enough non-`unstable` functionality because the actual source files will be more cluttered as well as the docsrs build - the docs thing (I've only just realized as typing) is pretty ugly.

**Reply by tcharding** [reply]
I'm missing how `--all-features` breaks anything, it only breaks semver compliance not actual code.

**Reply by apoelstra** [reply]
"breaking semver compliance" means that you are breaking actual code, because either (a) users are pinning exact versions of `primitives` and therefore are not compatible with each other, or (b) they are not pinning exact versions, and they will wind up pulling in a version of `primitives` that their code doesn't work with.

**Reply by apoelstra** [reply]
> * `rust-bitcoin` depends on `primitives` with `unstable` unconditionally enabled - gets all functionality and is not broken with `--all-features`.

Now anybody who depends on `bitcoin` cannot also depend on `primitives`, because cargo will enable the `unstable` feature which disables semver compliance.

**Comment by tcharding** [comment]
> Now anybody who depends on `bitcoin` cannot also depend on `primitives`, because cargo will enable the unstable feature which disables semver compliance.

Why would anyone depend on `primitives` if they depend on `bitcoin`? In my mind the sole reason for depending on `primitives` is so one does not have to depend on `bitcoin`.

> "breaking semver compliance" means that you are breaking actual code, because either (a) users are pinning exact versions of primitives and therefore are not compatible with each other, or (b) they are not pinning exact versions, and they will wind up pulling in a version of primitives that their code doesn't work with.

I tried to find an example but could not, how can this series of releases break dowstream crates?

We release, for example, `1.0.0` with:

```rust
    /// Adds `rhs` to self.
    #[cfg(feature = "unstable")]
    pub fn add(self, rhs: u32) -> u32 { self + rhs }
```

We then release `1.1.0` with (dummy example of breaking change):
```rust
    /// Adds `rhs` to self.
    pub fn add(self, rhs: usize) -> usize { self + rhs }
```

- Build of multiple deps causes `cargo` to build with version `1.1.0`
- Any downstream crate that does not use `unstable` gets the new function `add` - non-breaking, additive change.
- Any downstream crate that uses `unstable` gets broken in a semver non-compliant way, like we said they would.

I don't see how the non-`unstable` user ever gets broken. I'm suggesting that the code in `unstable` is correct its just the promise to be semver compliant that is violated if one _calls_ the unstable functions, not if one just builds them.

**Reply by apoelstra** [reply]
> Why would anyone depend on `primitives` if they depend on `bitcoin`? In my mind the sole reason for depending on `primitives` is so one does not have to depend on `bitcoin`.

Somebody trying to use `primitives` but who finds that they need to drag `bitcoin` in to handle missing functionality.

> * Any downstream crate that does not use `unstable` gets the new function `add` - non-breaking, additive change.

Ok, yes, if everyone is sure to only use `unstable` functions if they directly enable `unstable` (which the compiler/cargo will not help enforce) and if we are very careful to make sure that we only make breaking changes to things that are already marked `unstable`, then I believe there would be no breakage except to people who opt-in to `unstable`. Maybe this is manageable.

I would prefer a convention where we stick the suffix `_unstable` onto functions that we intend to break, and `Unstable` onto types that we intend to break. This would give us much stronger guardrails against accidental `unstable` usage, and also keep `--all-features` "pure".

**Reply by tcharding** [reply]
If we test in `primitives` CI that we can build without `unstable` then we cannot accidentally mess it up, right?

**Reply by apoelstra** [reply]
No, every dependent of `primitives` would need to be tested without `unstable`, something that may not be possible if anything else in their tree turns on `unstable`.

**Reply by tcharding** [reply]
Perhaps its time we maintain a branch on `miniscript` and a branch on `bitcoind-json-rpc` that tracks `bitcoin` master branch so we can prove things as we go.

**Reply by apoelstra** [reply]
We should do that, yes, but as long as those depend on `bitcoin` it will be impossible for them to depend on non-`unstable` `primitives`.

**Reply by tcharding** [reply]
Yes of course, I could be missing something but I'm confident non-`unstable` users will be okay. I'll find some crates in the wild to use to test it. `bdk` has some I think.

**Comment by tcharding** [comment]
For readers, @apoelstra and I discussed this offline and agreed to use an unstable marker on function names instead of a feature e.g., `foo__unstable`. For the PR that introduces this see: #3406

**Comment by tcharding** [comment]
Documentation tracked in #4855
